CSM Docs (#1996)
unoplatform/uno.extensions: doc/Learn/Tutorials/Markup/HowTo-CustomMarkupProject-Toolkit.md

---
uid: Learn.Tutorials.HowToCustomMarkupProjectToolkit
---

# Getting Started with Uno Toolkit

> In the previous session we learn how to [Create your own C# Markup](xref:Learn.Tutorials.HowToCreateMarkupProject) and how [Custom your own C# Markup - Learn how to change Style, Bindings, Templates and Template Selectors using C# Markup](xref:Learn.Tutorials.HowToCustomMarkupProject).

Reviewer: ```suggestion
In the previous session we learned how to [Create your own C# Markup](xref:Learn.Tutorials.HowToCreateMarkupProject) and how [Custom your own C# Markup - Learn how to change Style, Bindings, Templates and Template Selectors using C# Markup](xref:Learn.Tutorials.HowToCustomMarkupProject).
```

> Now we will check [Uno Toolkit](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/getting-started.html) with is a library that can be added to any new or existing Uno solution.

Reviewer: ```suggestion
Now we will check [Uno Toolkit](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/getting-started.html) which is a library that can be added to any new or existing Uno solution.
```

> # Getting Started with Uno Toolkit
> 
> In the previous session we learned how to [Create your own C# Markup](xref:Learn.Tutorials.HowToCreateMarkupProject) and how [Custom your own C# Markup - Learn how to change Style, Bindings, Templates and Template Selectors using C# Markup](xref:Learn.Tutorials.HowToCustomMarkupProject).
> 
> Now we will check [Uno Toolkit](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/getting-started.html) which is a library that can be added to any new or existing Uno solution.
> 
> For this sample you can use the how to [Create your own C# Markup with Toolkit](xref:Reference.Markup.HowTo-MarkupProjectToolkit)
> 
> For this sample we will cover this controls:
> 
> - [NavigationBar](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/NavigationBar.html)  - Represents a specialized app bar that provides layout for AppBarButton and navigation logic.
> 
> - [Chip](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/ChipAndChipGroup.html) - Chips are compact elements that represent an input, attribute, or action.

Reviewer 2: ```suggestion
# Getting Started with Uno Toolkit

In the previous session we learned how to [Create your own C# Markup](xref:Learn.Tutorials.HowToCreateMarkupProject) and how [Custom your own C# Markup - Learn how to change Style, Bindings, Templates, and Template Selectors using C# Markup](xref:Learn.Tutorials.HowToCustomMarkupProject).

Now we will check [Uno Toolkit](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/getting-started.html) which is a library that can be added to any new or existing Uno solution.

For this sample you can use the how to [Create your own C# Markup with Toolkit](xref:Reference.Markup.HowTo-MarkupProjectToolkit)

For this sample, we will cover these controls:

- [NavigationBar](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/NavigationBar.html)  - Represents a specialized app bar that provides layout for AppBarButton and navigation logic.

- [Chip](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/ChipAndChipGroup.html) - Chips are compact elements that represent an input, attribute, or action.
```

### Changing UI to have the [NavigationBar](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/NavigationBar.html)

- In the Shared Project open the file *MainPage.cs* and change the content to have the NavigationBar.

    # [**C# Markup**](#tab/cs)

    #### C# Markup

    To do so, create a new class file named SecondPage.
    Then open the class and replace the content for:

    ```csharp
    namespace MySampleToolkitProject;

    public sealed partial class SecondPage : Page
    {
	    public SecondPage()
	    {
		    this
			    .Background(ThemeResource.Get<Brush>("ApplicationPageBackgroundThemeBrush"))
			    .Content(new StackPanel()
			    .VerticalAlignment(VerticalAlignment.Center)
			    .HorizontalAlignment(HorizontalAlignment.Center)
			    .Children(
				    new NavigationBar().Content("Title Second Page"),
				    new Button()
					    .Content("Go to Main Page")
					    .Assign(out var navigationButton)
			    ));

		    navigationButton.Click += (s, e) =>
		    {
			    Frame.Navigate(typeof(MainPage));
		    };
	    }
    }

    ```

>     Now we can change the content of the MainPage.cs to have the NavigationBar.
> 
>     ```csharp
>     new NavigationBar().Content("Title Main Page")
>     ```
> 
>     And For have the controls to navigate for the SecondPage we can add a Button for do it.
> 
>     ```csharp
>     new Button()
> 		.Content("Go to Second Page")
> 		.Assign(out var navigationButton),
>     ```
> 
>     The main ideia here is to create a button that have Assign it self the output variable named in this case as navigationButton.
>     After that we need to add a Click EventHandler to add the Navigation using the Frame.Navigate. 
> 
>     > Notice how simple it is to create an action for the Button's Click event.

Reviewer 2: ```suggestion
    Now we can change the content of the MainPage.cs to have the NavigationBar.

    ```csharp
    new NavigationBar().Content("Title Main Page")
    ```

    And to have the controls to navigate for the SecondPage we can add a Button to do it.

    ```csharp
    new Button()
		.Content("Go to Second Page")
		.Assign(out var navigationButton),
    ```

    The main idea here is to create a button that assigns itself the output variable named in this case navigationButton.
    After that, we need to add a Click EventHandler to add the Navigation using the Frame.Navigate. 

    > Notice how simple it is to create an action for the Button's Click event.
```

    ```xml
    <Page x:Class="MySampleToolkitProjectXAML.SecondPage"
	    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
	    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
	    xmlns:local="using:MySampleToolkitProjectXAML"
	    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
	    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
	    mc:Ignorable="d"
	    xmlns:utu="using:Uno.Toolkit.UI"
	    Background="{ThemeResource ApplicationPageBackgroundThemeBrush}">

	    <StackPanel HorizontalAlignment="Center"
				    VerticalAlignment="Center">
		    <utu:NavigationBar Content="Title Main Page"/>
		    <Button x:Name="navigationButton" Content="Go to Second Page" Click="navigationButton_Click"/>
	    </StackPanel>
    </Page>
    ```

    And the SecondPage.cs

    ```csharp
    namespace MySampleToolkitProjectXAML;

    public sealed partial class SecondPage : Page
    {
	    public SecondPage()
	    {
		    this.InitializeComponent();
	    }
	    public void navigationButton_Click(object sender, RoutedEventArgs e)
	    {
		    Frame.Navigate(typeof(MainPage));
	    }
    }
    ```

>     Now we can change the content of the MainPage.xaml to have the NavigationBar.
> 
>     ```xaml
>     <utu:NavigationBar Content="Title Main Page"/>
>     ```
> 
>     And For have the controls to navigate for the SecondPage we can add a Button for do it.
> 
>     ```csharp
>     <Button 
> 			Content="Go to Second Page" 
> 			Click="navigationButton_Click"/>
>     ```
> 
>     # [**Full Code**](#tab/code)
> 
>     #### Full C# Markup code
>     
>     - Example of the complete code on the MainPage.cs, so you can follow along in your own project.

Reviewer 2: ```suggestion
    Now we can change the content of the MainPage.xaml to have the NavigationBar.

    ```xaml
    <utu:NavigationBar Content="Title Main Page"/>
    ```

    And to have the controls to navigate for the SecondPage we can add a Button to do it.

    ```csharp
    <Button 
			Content="Go to Second Page" 
			Click="navigationButton_Click"/>
    ```

    # [**Full Code**](#tab/code)

    #### Full C# Markup code
    
    - Example of the complete code on the MainPage.cs, so you can follow along in your own project.
```

    ```csharp
    namespace MySampleToolkitProject;

    public sealed partial class MainPage : Page
    {
	    public MainPage()
	    {
		    this
		        .Background(ThemeResource.Get<Brush>("ApplicationPageBackgroundThemeBrush"))
			    .Content(new StackPanel()
			    .VerticalAlignment(VerticalAlignment.Center)
			    .HorizontalAlignment(HorizontalAlignment.Center)
			    .Children(
				    new NavigationBar().Content("Title Main Page"),
				    new Button()
					    .Content("Go to Second Page")
					    .Assign(out var navigationButton)
			    ));

		    navigationButton.Click += (s, e) =>
		    {
			    Frame.Navigate(typeof(SecondPage));
		    };
	    }
    }
    ```

    #### Full XAML code

>     Look how the MainPage.xaml and the MainPage.xaml.cs will like like.

Reviewer 2: like like ?

	    <StackPanel HorizontalAlignment="Center"
				    VerticalAlignment="Center">
		    <utu:NavigationBar Content="Title Main Page"/>
		    <Button 
			    Content="Go to Second Page" 
			    Click="navigationButton_Click"/>
	    </StackPanel>
    </Page>

    ```

## [Chip](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/ChipAndChipGroup.html)

A Chip is a user interface component used to display a specific selection or action in a compact format.

Chips are often used to display short pieces of information such as tags, categories, filters, or selected options. They provide a compact and interactive visual representation of this information, allowing users to efficiently view and interact with it.

### Changing UI to have the Chip

- In the Shared Project open the file *MainPage.cs* and change the content to have the Chip.

    # [**C# Markup**](#tab/cs)

    #### C# Markup

    Chips can be displayed in lists, dropdown menus, action bars, or other areas of the user interface depending on the application's needs. They are highly flexible and adaptable, allowing you to use them according to your application's interaction flow.

>     For this case we will add the Chip to show some information on the UI.

Reviewer 2: ```suggestion
    For this case, we will add the Chip to show some information on the UI.
```

    ```csharp
    new ChipGroup()
		.Assign(out var chipGroup)
		.Items(
			new Chip()
				.Margin(5)
				.Content("Go to Second Page")
				.Background(new SolidColorBrush(Colors.LightBlue))
    			.Assign(out var navigationChip),
		)
    ```

    You can customize the appearance and behavior of Chips by setting properties such as Content, Icon, IsSelected, and handling associated events such as Click or Selected.

>     So lets add more Chip to see they in action.
>     The Second is for show how to custom the style of the chip.

Reviewer 2: ```suggestion
    So let's add more Chip to see them in action.
    The Second is to show how to customize the style of the chip.
```

    ```csharp
    new ChipGroup()
		.Assign(out var chipGroup)
		.Items(
            new Chip()
				.Margin(5)
				.Content("Chip 2")
				.Style(new Style<Chip>()
							.Setters( s => s.Foreground(new SolidColorBrush(Colors.Red)))
					),
		)
    ```

>     And the third one is for show how to handle the Checked and the Unchecked events.

Reviewer 2: ```suggestion
    And the third one is to show how to handle the Checked and the Unchecked events.
```

    ```csharp
    new ChipGroup()
		.Assign(out var chipGroup)
		.Items(
			
			new Chip()
				.Margin(5)
				.Content("Chip 3")
				.Assign(out var chipElement)
		)
    ```

>     And we need to add the event handlers  on our code.
>     Notice that we are using the Assign to have access to the chipElement in other places.

Reviewer 2: ```suggestion
    And we need to add the event handlers to our code.
    Notice that we are using the Assign to have access to the chipElement in other places.
```

    
    ```csharp
    
	chipElement.Checked += (sender, e) =>
	{
		if (sender is Chip chip)
		{
			chip.FontSize(18);
		}
	};
	chipElement.Unchecked += (sender, e) =>
	{
		if (sender is Chip chip)
		{
			chip.FontSize(14);
		}
	};
    ```

>     > Notice that we are able to use Event handlers and use for many other propous.
> 
>     # [**XAML**](#tab/cli)
>     
>     #### XAML
>     
>     For this case we will add the Chip to show some information on the UI.

Reviewer 2: ```suggestion
    > Notice that we are able to use Event handlers and use them for many other purposes.

    # [**XAML**](#tab/cli)
    
    #### XAML
    
    For this case, we will add the Chip to show some information on the UI.
```

    ```xml
    <utu:ChipGroup x:Name="chipGroup">
		<utu:Chip Margin="5" Content="Chip 1" Background="LightBlue" />
	</utu:ChipGroup>
    ```

>     You can customize the appearance and behavior of Chips by setting properties such as Content, Icon, IsSelected, and handling associated events such as Click or Selected.
> 
>     So lets add more Chip to see they in action.
>     The Second is for show how to custom the style of the chip.

Reviewer 2: ```suggestion
    You can customize the appearance and behavior of Chips by setting properties such as Content, Icon, IsSelected, and handling associated events such as Click or Selected.

    So let's add more Chip to see them in action.
    The Second is to show how to customize the style of the chip.
```

    ```xml
    <utu:ChipGroup x:Name="chipGroup">
		<utu:Chip Margin="5" Content="Chip 2">
			<utu:Chip.Style>
				<Style TargetType="utu:Chip">
					<Setter Property="Foreground" Value="Red" />
				</Style>
			</utu:Chip.Style>
		</utu:Chip>
	</utu:ChipGroup>
    ```

>     And the third one is for show how to handle the Checked and the Unchecked events.

Reviewer 2: ```suggestion
    And the third one is to show how to handle the Checked and the Unchecked events.
```

    ```xml
    <utu:ChipGroup x:Name="chipGroup">
		<utu:Chip Margin="5" Content="Chip 3" x:Name="chipElement" Checked="chip_Checked" Unchecked="chip_Unchecked"/>
	</utu:ChipGroup>
    ```

>     And we need to add the event handlers  on our code.

Reviewer 2: ```suggestion
    And we need to add the event handlers on our code.
```

		
		        <utu:ChipGroup x:Name="chipGroup">
			        <utu:Chip Margin="5" Content="Chip 1" Background="LightBlue" />
			        <utu:Chip Margin="5" Content="Chip 2">
				        <utu:Chip.Style>
					        <Style TargetType="utu:Chip">
						        <Setter Property="Foreground" Value="Red" />
					        </Style>
				        </utu:Chip.Style>
			        </utu:Chip>
			        <utu:Chip Margin="5" Content="Chip 3" x:Name="chipElement" Checked="chip_Checked" Unchecked="chip_Unchecked"/>

		        </utu:ChipGroup>
	        </StackPanel>
        </StackPanel>
    </Page>
    ```

    - MainPage.xaml.cs

    ```csharp
    namespace MySampleToolkitProjectXAML;

    public sealed partial class MainPage : Page
    {
	    public MainPage()
	    {
		    this.InitializeComponent();
	    }
	    public void navigationButton_Click(object sender, RoutedEventArgs e)
	    {
		    Frame.Navigate(typeof(SecondPage));
	    }
	    public void chip_Unchecked(object sender, RoutedEventArgs e)
	    {
		    chipElement.FontSize = 14;
	    }
	    public void chip_Checked(object sender, RoutedEventArgs e)
	    {
		    chipElement.FontSize = 18;
	    }
	
    }

    ```

> ## Try it yourself
> 
> Now try to change your MainPage to have different layout and test other attributes and elements.
> 
> In this Tutorial we add the NavigationBar and some Chip to the UI.
> 
> But the Uno Toolkit has many other Controls as:
> 
> - [AutoLayout](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/AutoLayoutControl.html)
> - [Cards](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/CardAndCardContentControl.html)
> - [Chip and ChipGroup](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/ChipAndChipGroup.html)
> - [DrawerControl](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/DrawerControl.html)
> - [DrawerFlyoutPresenter](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/DrawerFlyoutPresenter.html)
> - [LoadingView](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/LoadingView.html)
> - [NavigationBar](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/NavigationBar.html)
> - [SafeArea](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/SafeArea.html)
> - [TabBar and TabBarItem](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/TabBarAndTabBarItem.html)
> 
> > Try to add another control as a learning exercise.
> 
> ## Next Steps
> 
> - [Custom your own C# Markup - Learn how to change Visual States and User Controls](xref:Learn.Tutorials.HowToCustomMarkupProjectVisualStates)
> - [Custom your own C# Markup - Learn how to Change the Theme](xref:Learn.Tutorials.HowToCustomMarkupProjectTheme)
> - [Custom your own C# Markup - Learn how to use MVUX](xref:Learn.Tutorials.HowToCustomMarkupProjectMVUX)

Reviewer 2: ```suggestion
## Try it yourself

Now try to change your MainPage to have a different layout and test other attributes and elements.

In this Tutorial, we add the NavigationBar and some Chip to the UI.

But the Uno Toolkit has many other Controls as:

- [AutoLayout](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/AutoLayoutControl.html)
- [Cards](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/CardAndCardContentControl.html)
- [Chip and ChipGroup](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/ChipAndChipGroup.html)
- [DrawerControl](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/DrawerControl.html)
- [DrawerFlyoutPresenter](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/DrawerFlyoutPresenter.html)
- [LoadingView](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/LoadingView.html)
- [NavigationBar](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/NavigationBar.html)
- [SafeArea](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/SafeArea.html)
- [TabBar and TabBarItem](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/TabBarAndTabBarItem.html)

> Try to add another control as a learning exercise.

## Next Steps

- [Custom your own C# Markup - Learn how to change Visual States and User Controls](xref:Learn.Tutorials.HowToCustomMarkupProjectVisualStates)
- [Custom your own C# Markup - Learn how to Change the Theme](xref:Learn.Tutorials.HowToCustomMarkupProjectTheme)
- [Custom your own C# Markup - Learn how to use MVUX](xref:Learn.Tutorials.HowToCustomMarkupProjectMVUX)
```
